libraries/AP_HAL_PX4/I2CDevice.cpp

```diff
@@ -67,6 +107,122 @@ bool I2CDevice::read_registers_multiple(uint8_t first_reg, uint8_t *recv,
     return false;
 }
 
+/*
+  per-bus i2c callback thread
+*/
+void *I2CDevice::i2c_thread(void *arg)
+{
+    struct bus_info *binfo = (struct bus_info *)arg;
+    while (!_px4_thread_should_exit) {
+        uint64_t now = AP_HAL::micros64();
+        uint64_t next_needed = 0;
+        struct callback_info *callback;
+
+        // find a callback to run
+        for (callback = binfo->callbacks; callback; callback = callback->next) {
+            if (now >= callback->next_usec) {
+                while (now >= callback->next_usec) {
+                    callback->next_usec += callback->period_usec;
+                }
+                // call it with semaphore held
+                if (binfo->semaphore.take(0)) {
+                    callback->cb();
+                    binfo->semaphore.give();
+                }
+            }
+            if (next_needed == 0 ||
+                callback->next_usec < next_needed) {
+                next_needed = callback->next_usec;
+            }
+        }
+
+        // delay for at most 50ms, to handle newly added callbacks
+        now = AP_HAL::micros64();
+        uint32_t delay = 50000;
+        if (next_needed > now && next_needed - now < delay) {
+            delay = next_needed - now;
+        }
+        hal.scheduler->delay_microseconds(delay);
+    }
+    return nullptr;
+}
+    
+
+/*
+  find or create thread for this bus
+*/
+I2CDevice::bus_info *I2CDevice::find_thread(void)
+{
+    if (_busnum >= num_buses) {
+        return nullptr;
+    }
+    struct bus_info &binfo = businfo[_busnum];
+    if (binfo.thread_started) {
+        return &binfo;
+    }
+
+    binfo.bus = _busnum;
+    binfo.thread_started = true;
+    
+    pthread_attr_t thread_attr;
+    struct sched_param param;
+    
+    pthread_attr_init(&thread_attr);
+    pthread_attr_setstacksize(&thread_attr, 1024);
+    
+    param.sched_priority = APM_I2C_PRIORITY;
+    (void)pthread_attr_setschedparam(&thread_attr, &param);
+    pthread_attr_setschedpolicy(&thread_attr, SCHED_FIFO);
+    
+    pthread_create(&binfo.thread_ctx, &thread_attr, &I2CDevice::i2c_thread, &binfo);
```

Comment:
Humn... maybe having a Thread implementation just like in the Linux layer would be good to reduce the boilerplate code to create/start/stop a thread.